Implement dangling indices API

pugnascotia opened the pull request:
Part of #48366. Implement an API for managing dangling indices.

~~I've been staring at this code for a while so I'd really appreciate some feedback. Note that I haven't (yet) implemented wildcard restores, or written API documentation.~~

pugnascotia committed:
Add an API to fetch dangling indices

pugnascotia committed:
Refactor list dangling indices to call all nodes

pugnascotia committed:
WIP - working on restore API

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Wow, restores seem to work

pugnascotia committed:
Use a better REST model for restores

pugnascotia committed:
Test fix

pugnascotia committed:
WIP - working on a delete endpoint

pugnascotia committed:
Delete seems to work. Wow.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Polishing

pugnascotia committed:
Check data loss flag after validating an index is dangling

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Check data loss flag when restoring dangling indices

pugnascotia committed:
Ensure nodes are removed before deleting indices

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Add lots of Javadoc

pugnascotia committed:
Add missing license info

pugnascotia committed:
Checkstyle

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Recovery)

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

ywelsch reviewed (changes requested):
Thanks for working on this. I've left some comments and items to discuss.

We should break this into smaller reviewable chunks once we agree on the direction.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Rename restore to import, uppercase UUID refs

Refer to "importing" dangling indices rather than "restoring" them. Also
keeping UUID uppercased in various places, instead of camel-cased.

pugnascotia committed:
Add node_name and creation_date fields to DanglingIndexInfo

pugnascotia committed:
Convert danling index API request bodies to URL params

pugnascotia committed:
Cleanups

pugnascotia commented on .eclipseformat.xml:
```diff
@@ -62,7 +62,7 @@
         <setting id="org.eclipse.jdt.core.formatter.insert_new_line_after_annotation_on_method" value="insert"/>
         <setting id="org.eclipse.jdt.core.formatter.insert_space_after_opening_paren_in_switch" value="do not insert"/>
-        <setting id="org.eclipse.jdt.core.formatter.alignment_for_parameterized_type_references" value="0"/>
+        <setting id="org.eclipse.jdt.core.formatter.alignment_for_parameterized_type_references" value="48"/>
```
This is now in master and will disappear from the diffs.

pugnascotia commented:
@ywelsch thank you for the view. As you may have gathered, I didn't entirely know what I was doing in places 🙈 

I've pushed some changes to address the following:

   - REST request bodies replaced with query parameters
   - Standardised on `index_uuid` in the JSON specs, and `indexUUID` in Java.
   - Renamed all references to "restoring" a dangling index to "importing".
   - Use the `GENERIC` thread pool for deletes (I'm guessing, please advise)

Regarding the bigger questions that you asked:

   - I'm using the graveyard at Dave's suggestion. I'd be very happy for further guidance here.
   - I'm querying all the dangling indices on a delete request in order to validate the supplied index UUID. I'll look at reducing the data being shuffled between nodes once we've settled the other aspects.
   - I had intended to implement the API, then go back and refactor the scan-on-cluster-update behaviour afterwards.

@DaveCTurner did you have any thoughts here, either on the code changes or regarding Yannick's comments?

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/TransportDeleteDanglingIndexAction.java:
```diff
@@ -0,0 +1,225 @@
+/**
+ * Implements the deletion of a dangling index. When handling a {@link DeleteDanglingIndexAction},
+ * this class first checks that such a dangling index exists. It then submits a cluster state update
+ * to add the index to the index graveyard.
```
I think we do indeed want to put it in the index graveyard, so that we don't need to be sure that all the nodes on which the index still exists are in the cluster at the time. I think it would be surprising to delete a dangling index, get a `200 OK` response, and yet the dangling index persists because of a network blip or a node that's still starting.

I don't see a problem with involving the master here - these updates will be pretty low-volume.

I do, however, see that this will make it tricky to completely remove the index graveyard in future. I think that's ok, we can still stop using the index graveyard for indices that we're certain were deleted.

pugnascotia commented:
Oh, I added the node name and creation date into the list response payload too.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/DeleteDanglingIndexResponse.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder.startObject().field("status", "ok").endObject();
```
I was trying to keep this the same as something, but I can't remember what. I checked what we send back when we delete a regular index, and replaced this with `acknowledged: true`. This isn't strictly necessary, but I feel that it's nice to give feedback e.g. to `curl` users.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/ListDanglingIndicesResponse.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public RestStatus status() {
+        return this.hasFailures() ? RestStatus.SERVICE_UNAVAILABLE : RestStatus.OK;
```
I was trying to differentiate between "the server couldn't handle the request", and "some nodes couldn't handle the request". What do you think, is that overkill?

pugnascotia committed:
Improve NodeDanglingIndicesResponse serialisation

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/package-info.java:
```diff
@@ -0,0 +1,37 @@
+ * <p>The classes in this package form an API for managing dangling
+ * indices, allowing them to be listed, imported or deleted.
+ */
+package org.elasticsearch.action.admin.indices.dangling;
```
💯 for package-level javadocs 😀

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -191,27 +192,49 @@ public void testDanglingIndicesStripAliases() throws Exception {
             when(clusterServiceMock.getSettings()).thenReturn(allocateSettings);
 
-            final DanglingIndicesState danglingIndicesState = new DanglingIndicesState(
+            DanglingIndicesState danglingIndicesState = new DanglingIndicesState(
```
Still `final`?

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -191,27 +192,49 @@ public void testDanglingIndicesStripAliases() throws Exception {
                 metaStateService,
-                localAllocateDangledIndices,
-                clusterServiceMock
+                localAllocateDangledIndices, clusterServiceMock
```
nit: unnecessary whitespace change

DaveCTurner reviewed (commented):
It's mostly heading in the right direction, but I left a selection of smaller comments. There's still a few things from Yannick's earlier review to address.

pugnascotia committed:
Mark APIs as stable

pugnascotia committed:
Make DeleteDanglingIndexRequest immutable

pugnascotia committed:
Make ImportDanglingIndexRequest immutable

pugnascotia committed:
Remove redundant validation

pugnascotia committed:
Replace hand-rolled toString() with generated

pugnascotia committed:
Review feedback

pugnascotia committed:
Aggregate list dangling indices rest response by ID

pugnascotia committed:
Remove blocking call from delete danling index

pugnascotia committed:
Remove blocking call from import danlging index

pugnascotia committed:
Pick the latest metadata when importing a dangling index

pugnascotia committed:
Tweaks

pugnascotia committed:
Refactor to use async requests

Refactor inter-node requests to be async, and introduce a find action so
that the import and delete dangling index commands don't have to pull
too much data over the network.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Fix unstable test

pugnascotia committed:
Nest restartRandomDataNode calls and delete restartRandomDataNodes(int,RestartCallback)

pugnascotia committed:
Polishing

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/ImportDanglingIndexRequest.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public ActionRequestValidationException validate() {
+        if (this.indexUUID == null || this.indexUUID.isEmpty()) {
```
Yes, but IMO it would indicate an error somewhere and I'd rather fail fast.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/ListDanglingIndicesResponse.java:
```diff
@@ -0,0 +1,99 @@
+        }
+        builder.endArray();
+
+        if (this.hasFailures()) {
```
I couldn't see how that method would be invoked for the new API, but I've swapped my implementation for a call to `buildNodesHeader`.

DaveCTurner replied:
Oh I see, I think we need this:

```diff
diff --git a/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/dangling/RestListDanglingIndicesAction.java b/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/dangling/RestListDanglingIndicesAction.java
index 55370801e64..4604b558668 100644
--- a/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/dangling/RestListDanglingIndicesAction.java
+++ b/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/dangling/RestListDanglingIndicesAction.java
@@ -43,6 +43,6 @@ public class RestListDanglingIndicesAction extends BaseRestHandler {
     @Override
     public BaseRestHandler.RestChannelConsumer prepareRequest(final RestRequest request, NodeClient client) throws IOException {
         final ListDanglingIndicesRequest danglingIndicesRequest = new ListDanglingIndicesRequest();
-        return channel -> client.admin().cluster().listDanglingIndices(danglingIndicesRequest, new RestStatusToXContentListener<>(channel));
+        return channel -> client.admin().cluster().listDanglingIndices(danglingIndicesRequest, new RestActions.NodesResponseRestListener<>(channel));
     }
 }
```

pugnascotia committed:
Log level tweak

pugnascotia committed:
Fix the very broken dangling indices REST API

pugnascotia committed:
Write REST tests for dangling indices

pugnascotia committed:
JavaDoc

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Updates after merge, move REST IT

Changes in master required updates to the new REST controllers. Also
move `DanglingIndicesRestIT` to the `qa:smoke-test-http` project,
because it's a slightly better home than the `modules:transport-netty4`
project.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/ListDanglingIndicesResponse.java:
```diff
@@ -0,0 +1,98 @@
+                    indexMetaData.getIndex().getName(),
+                    indexMetaData.getIndexUUID()
+                );
+                danglingIndexInfo.toXContent(builder, params);
```
I've reworked the response shape, and added a REST version of the integration test. It's currently living under `qa:smoke-test-http`, which I appreciate isn't ideal. At least the REST code is now tested.

pugnascotia commented:
@ywelsch @DaveCTurner I *think* the only outstanding point here now is the refactoring of how dangling indices are detected, which I would still prefer to do as a dedicated PR. Would you mind taking another look please?

pugnascotia committed:
Checkstyle

pugnascotia committed:
Satify :server:loggerUsageCheck

pugnascotia committed:
Test fixes

pugnascotia commented:
@ywelsch @DaveCTurner still looking for feedback / further direction on this PR.

DaveCTurner reviewed (changes requested):
I left a handful of minor comments; the main outstanding issue is about when we scan for dangling indices.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Rename NodeDanglingIndicesResponse to NodeListDanglingIndicesResponse

pugnascotia committed:
Add a comment in NodeFindDanglingIndexResponse

pugnascotia committed:
Rename NodeDanglingIndicesRequest to NodeListDanglingIndicesRequest

pugnascotia committed:
Logging tweaks

pugnascotia committed:
Use NodeClient in TransportDeleteDanglingIndexAction

pugnascotia committed:
Use NodeClient in TransportImportDanglingIndexAction

pugnascotia committed:
Only poll dangling indices if setting is enabled

`gateway.auto_import_dangling_indices` used to only control whether
dangling indices would be automatically imported when found. This commit
expands this setting's role so that when it's disabled,
DanglingIndicesState won't listen for cluster state changes at all.
Insetad, the new `/_dangling` API must be used.

pugnascotia committed:
Checkstyle

pugnascotia commented:
@DaveCTurner I've addressed your latest feedback, in particular the code no longer polls for dangling indices when `gateway.auto_import_dangling_indices` is `false`.

DaveCTurner reviewed (changes requested):
Thanks @pugnascotia I think this works well now. I have given it a more thorough review and left some small suggestions.

I ran out of steam before properly looking at `DanglingIndicesIT` but I've been through everything else.

ywelsch commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/NodeFindDanglingIndexResponse.java:
```diff
@@ -0,0 +1,62 @@
+     * UUID if the situation is really crazy, though perhaps this is more likely
+     * when collating responses from different nodes.
+     */
+    private final List<IndexMetaData> danglingIndexMetaData;
```
Can we avoid sending the full index metadata here? This object is potentially very big, and we are only interested in two small fields from that object.

DaveCTurner replied:
Note that we're only sending the metadata for a single index here. Although it might be quite big it shouldn't be unmanageably so.

There's a couple of options for trimming it down further: we could distinguish the `find` action needed for deletion from the one needed for import, or else we could share just the metadata version here and have the coordinating node pick one node and tells it to send its metadata to the master. Either way we have to add another transport action, and I don't think the savings are really worth it.

ywelsch replied:
ok, following your suggestion here

ywelsch commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/TransportDeleteDanglingIndexAction.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    private ClusterState deleteDanglingIndex(ClusterState currentState, Index indexToDelete) {
+        final MetaData meta = currentState.metaData();
```
do we need to double-check here that the index does not exist as proper index in the cluster state?

ywelsch commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/TransportDeleteDanglingIndexAction.java:
```diff
@@ -0,0 +1,207 @@
+
+        final IndexGraveyard.Builder graveyardBuilder = IndexGraveyard.builder(metaDataBuilder.indexGraveyard());
+
+        final IndexGraveyard newGraveyard = graveyardBuilder.addTombstone(indexToDelete).build(settings);
```
By just adding the index to the graveyard, how is that making the nodes actually delete the index when applying the CS update? I might be missing something here but can't find the logic for it.

DaveCTurner replied:
I think there's a testing gap here. We're checking that when we delete the index it no longer appears as a dangling index, but not actually checking that it is removed from disk. I think this means if we delete another couple of indices to phase this one out of the graveyard then we would see it reappear.

pugnascotia committed:
Test refactoring

pugnascotia committed:
Static import

pugnascotia committed:
More test refactoring

pugnascotia committed:
Expect dangling index allocation to be immediate

pugnascotia committed:
Check dangling index UUID in REST test

pugnascotia committed:
Rename variables

pugnascotia committed:
Use AcknowledgedResponse where possible

pugnascotia committed:
Simplify DanglingIndexInfo

pugnascotia committed:
Tweaks

pugnascotia committed:
Don't pass around IndexMetaData for deletes

Refactor dangling index classes into sub-packages. Rework the find
action to return a list of `DanglingIndexInfo`. Introduce a new find
action that still returns `IndexMetaData` since importing a dangling
index requires this.

pugnascotia committed:
Add optional filter to list requests

Make it possible to filter list requests by index UUID, which means that
the find actions can be consolidated once again.

pugnascotia committed:
Add check for index UUID in cluster state

Before performing a dangling index delete, first check that the dangling
index UUID doesn't exist in the cluster state.

pugnascotia committed:
Support timeout params in import and delete commands

pugnascotia committed:
Change import response type again

Use a tailored response type for importing a danging index, but include
"accepted" in the response payload rather than "acknowledged".

Also add unit tests for the list action's aggregation logic.

pugnascotia committed:
Tweaks

pugnascotia committed:
Remove find dangling index from cluster admin client

pugnascotia commented:
@ywelsch @DaveCTurner thank you again for the feedback, I believe I've addressed most of it.

The main exception is how to ensure dangling index files are deleted. I would welcome any guidance you may have here.

Also, I'm unsure about the suggestions around removing `@ClusterScope`. Right now, the test class `DanglingIndicesIT` starts clusters of varying sizes, with a variety of settings. It may be possible to consolidate this, but I haven't yet attempted it.

Notes on changes since the last view:

* Reduce deplicated code in ITs
* Accept timeout parameters in import and delete requests
* Move new classes around into sub-packages. IntelliJ wasn't happy with a sub-package name of `import`, which I assume is reserved, so we have `import_index` in this case.
* Cleap up class hierarchies - there were some oddities left over from previous iterations
* Ensure no index with the dangling UUID exists in the cluster state before deletion
* Tweak the list action so that its response can be filtered by UUID. This allows the delete action to use it instead of the find action, and cuts down the data being transferred. The find action still returns the full metadata because the code that imports dangling indices expects to be passed it.
* As the code now stands, it's less easy to extract common code from the import and delete transport actions. I tried but decided it wasn't worth it.
* Added unit tests for the list aggregation code.
* Removed the find action from the cluster admin API

pugnascotia committed:
Checkstyle

pugnascotia committed:
Fix DanglingIndicesStateTests

Replace a test, which was checking allocation behaviour when dangling
indices were detected, with a test that checks that no cluster state
listener is registered when the detection setting is disabled.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Fix checkstyle

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

pugnascotia commented:
@ywelsch @DaveCTurner do you have time for another look at this PR after my [last round of updates](https://github.com/elastic/elasticsearch/pull/50920#issuecomment-593912077)?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

DaveCTurner commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java:
```diff
@@ -0,0 +1,292 @@
+ *
+ * @see org.elasticsearch.action.admin.indices.dangling
+ */
+@ClusterScope(numDataNodes = 0, scope = ESIntegTestCase.Scope.TEST, autoManageMasterNodes = false)
```
Ok I looked into this and now think I prefer this to the alternative I suggested in which we configure `SETTING_MAX_TOMBSTONES` to 1 in all the tests and then work our way around that, so no action required here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -853,4 +853,41 @@ public static String padStart(String s, int minimumLength, char c) {
+     * @throws NullPointerException if {@code obj} is {@code null}
+     * @throws IllegalArgumentException if {@code obj.isEmpty()} returns {@code true}
+     */
+    public static String requireNonEmpty(String obj, String message) {
```
Still outstanding -- let's remove this and just use `Objects.requireNonNull`.

pugnascotia replied:
> an empty UUID string is kinda valid

I'm not sure I agree with this statement. The API is designed to import a specific UUID. You can't even POST to the REST API without passing a UUID, because the REST layer infers that you ought to sending a GET.

But, putting that aside, it's highly unlikely that anything is going to call the transport API but our code, so I've swapped the calls out as suggested.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:
```diff
@@ -348,4 +350,80 @@ public static int sortAndDedup(final BytesRefArray bytes, final int[] indices) {
+     * @param <T2> the type that the returned list contains
+     * @return a list containing the result of applying the <code>mapper</code> function to each element in <code>list</code>.
+     */
+    public static <T1, T2> List<T2> map(List<T1> list, Function<T1, T2> mapper) {
```
This is still outstanding -- let's remove these and inline them at their few callers.

pugnascotia replied:
Inlined.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -161,11 +192,21 @@ void findNewAndAddDanglingIndices(final MetaData metaData) {
+                            indexMetaData.getIndex()
+                        );
+                    } else {
+                        logger.info(
```
This is still outstanding -- let's not log things here.

pugnascotia replied:
Removed.

DaveCTurner commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java:
```diff
@@ -0,0 +1,291 @@
+                + "  \"settings\": {"
+                + "    \"index\": {"
+                + "      \"number_of_shards\": 1,"
+                + "      \"number_of_replicas\": 2"
```
I think we also need to set `index.routing.allocation.total_shards_per_node: 1` to ensure that every node gets a copy of the shard.

DaveCTurner commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java:
```diff
@@ -0,0 +1,291 @@
+        final AtomicReference<String> stoppedNodeName = new AtomicReference<>();
+
+        // Restart node, deleting the index in its absence, so that there is a dangling index to recover
+        internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
```
Are you sure that the dangling index metadata was written to disk by this point? I think it happens asynchronously now.

pugnascotia replied:
I'm not sure - do you know how I would check? I had assumed that the tests would fail if the metadata wasn't written, because the create test explicitly checks what the dangling list API returns, and the import and delete tests both perform list calls before their respective actions.

DaveCTurner replied:
Yes I think these tests would fail if the metadata weren't written; I didn't see them fail but it might be pretty rare and I didn't try very hard to make it less rare, but nor could I see a reason why the metadata is definitely written by this point.

I don't think it's enough to list the dangling indices, or at least I don't see how this guarantees anything about the persistence of metadata of the non-dangling indices.

I was going to suggest adding a method to `PersistedState` but then I see that we already have one, see `GatewayMetaState#allPendingAsyncStatesWritten`. It should be enough to `assertBusy` that this returns true on all nodes.

DaveCTurner replied:
I see that you added an `assertBusy()` on `GatewayMetaState#allPendingAsyncStatesWritten` below. In fact I think we need to wait for that _before_ we restart the node, not after.

DaveCTurner commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java:
```diff
@@ -0,0 +1,291 @@
+        final Response deleteResponse = restClient.performRequest(deleteRequest);
+        assertThat(deleteResponse.getStatusLine().getStatusCode(), equalTo(ACCEPTED.getStatus()));
+
+        assertBusy(() -> assertThat("Expected dangling index to be deleted", listDanglingIndexIds(), hasSize(0)));
```
We can demonstrate that the dangling index was not properly deleted by adding this:

```
        createIndex("additional");
        deleteIndex("additional");
        assertThat(listDanglingIndexIds(), empty());
```

The creation/deletion of this extra index pushes the tombstone out of the graveyard which resurrects it.

(gruesome nomenclature isn't it? :grin:)

pugnascotia replied:
Well that worked, in the sense that it showed that the index isn't being deleted off disk. I'll see if I can figure out how to delete it properly, but if you have any pointers, that'd be great.

DaveCTurner replied:
One idea is for `ClusterChangedEvent#indicesDeletedFromClusterState` to consider added tombstones as an indication that an index was deleted as well as removed index metadata.

pugnascotia replied:
@DaveCTurner see commit d02c5ce3113. The tests now passes, but I'm really not sure about the changes. Can you please take a look?

DaveCTurner replied:
Yep, looks like what we want. Suggest using a `Set<Index>` rather than a list and a membership check but otherwise fine.

pugnascotia replied:
I switched to a set in b94efdf802b.

DaveCTurner reviewed (commented):
Thanks @pugnascotia, I withdrew one of my comments about using a nonempty cluster throughout since it turns out to be worse than the alternatives.

I also suggested a way to detect that the index isn't yet really being deleted.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Checkstyle

pugnascotia committed:
Dangling index deletes should remove data from disk

Attempt to ensure that when a dangling index is explicitly deleted, the
data on disk is also removed. While the end-to-end REST test now passes,
I'm really not sure about the changes to the server.

pugnascotia committed:
Checkstyle

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Fix REST API validation failures

pugnascotia committed:
Review feedback

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

DaveCTurner commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java:
```diff
@@ -0,0 +1,312 @@
+     * that each index has a primary or replica shard on every node, and if
+     * a node is stopped prematurely, this assumption is broken.
+     *
+     * @return a mapping from each createad index name to its UUID
```
Typo

```suggestion
     * @return a mapping from each created index name to its UUID
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        boolean noExceptionThrown = false;
+        try {
+            client().admin().cluster().importDanglingIndex(request).actionGet();
+            noExceptionThrown = true;
```
Could we `expectThrows(IllegalArgumentException.class, ...)` here instead of this?

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        try {
+            client().admin().cluster().importDanglingIndex(request).actionGet();
+        } catch (Exception e) {
+            caughtException = e;
```
Could we `expectThrows()` here?

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+
+        client().admin().cluster().deleteDanglingIndex(new DeleteDanglingIndexRequest(danglingIndexUUID, true)).actionGet();
+
+        assertThat(listDanglingIndices(), is(empty()));
```
WDYT about making the same fix to this test as the REST test to verify that the index really did get deleted?

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+            .deleteDanglingIndex(new DeleteDanglingIndexRequest(danglingIndices.get().get(0).getIndexUUID(), true))
+            .actionGet();
+
+        assertBusy(() -> assertThat(listDanglingIndices(), empty()));
```
Also here -- WDYT about making the same fix to this test as the REST test to verify that the index really did get deleted?

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        try {
+            client().admin().cluster().deleteDanglingIndex(new DeleteDanglingIndexRequest(danglingIndexUUID, false)).actionGet();
+        } catch (Exception e) {
+            caughtException = e;
```
Suggest `expectThrows` here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -134,24 +153,37 @@ void cleanupAllocatedDangledIndices(Metadata metadata) {
      * as dangled already.
      */
-    Map<Index, IndexMetadata> findNewDanglingIndices(final Metadata metadata) {
+    public Map<Index, IndexMetadata> findNewDanglingIndices(Map<Index, IndexMetadata> existingDanglingIndices, final Metadata metadata) {
```
Should probably have a unit test that passes something other than `emptyMap()` to this new parameter.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+    /**
+     * Check that when dangling indices are discovered, then they can be listed.
+     */
+    public void testDanglingIndicesCanBeListed() throws Exception {
```
Should we also test that this works if auto-import is enabled? You can make an un-importable dangling index by deleting it and creating another with the same name while the node is away.

DaveCTurner commented on server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java:
```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+     * other will be considered dangling, and can therefore be listed and
+     * deleted through the API
+     */
+    public void testDanglingIndexCanBeDeleted() throws Exception {
```
Should we also check that this works if auto-import is enabled?

DaveCTurner reviewed (commented):
Looking good @pugnascotia, I have been over this from top to bottom and left a few very small comments and questions.

pugnascotia committed:
Address review feedback

pugnascotia committed:
More tweaks

pugnascotia committed:
More tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Fix imports

pugnascotia commented:
@DaveCTurner thank you again, I believe I've addressed all your feedback.

Adding tests for the auto-import cases revealed that I needed to be more careful when checking for the presence of an index in the cluster metadata or graveyard - I had been doing it by name, as code elsewhere does, but actually it has to be done by UUID.

I also had to rework the dangling index detection code to tease apart detection vs allocation, so that in the auto-import case, the new API can still be used to list dangling indices that cannot be auto-imported.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

DaveCTurner committed:
Merge branch 'master' into 48366-dangling-indices-api

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -124,7 +154,7 @@ public void testDanglingProcessing() throws Exception {
             int numberOfChecks = randomIntBetween(1, 10);
             for (int i = 0; i < numberOfChecks; i++) {
-                Map<Index, IndexMetadata> newDanglingIndices = danglingState.findNewDanglingIndices(metadata);
+                Map<Index, IndexMetadata> newDanglingIndices = danglingState.findNewDanglingIndices(emptyMap(), metadata);
```
I don't think this is what happens when auto-import is enabled as it is here; instead we update the `danglingIndices` field and pass that in for subsequent calls.

pugnascotia replied:
Re: this and the other comment on this file...this is a weird test. The `findNewDanglingIndices` method doesn't modify anything, and yet it's being called in a loop. It makes more sense to call `findNewAndAddDanglingIndices` in a loop (which the test also does). So I feel like maybe the test need rewritten? What do you think?

DaveCTurner replied:
Good point, the calls to `findNewAndAddDanglingIndices` do what I suggest too. I'm ok with this as it is.

DaveCTurner commented on server/src/test/java/org/elasticsearch/gateway/DanglingIndicesStateTests.java:
```diff
@@ -142,7 +172,7 @@ public void testDanglingProcessing() throws Exception {
             // check that several runs when in the metadata, but not cleaned yet, still keeps dangled
             for (int i = 0; i < numberOfChecks; i++) {
-                Map<Index, IndexMetadata> newDanglingIndices = danglingState.findNewDanglingIndices(metadata);
+                Map<Index, IndexMetadata> newDanglingIndices = danglingState.findNewDanglingIndices(emptyMap(), metadata);
```
Similarly here, we should be using the previous result rather than `emptyMap()`.

DaveCTurner reviewed (commented):
Thanks @pugnascotia, I left a handful of small comments. Apart from that I'm 100% ok with the new APIs and all that infrastructure. I'll need another opinion on the changes to the existing code as I'm finding it very hard to follow all the different states that we can be in while deleting an index.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -134,38 +152,51 @@ void cleanupAllocatedDangledIndices(Metadata metadata) {
+                // dangling index files being deleted, so we don't expect to encounter a dangling index and a tombstone here when
+                // everything is working normally.
+                if (graveyard.containsIndex(index)) {
+                    logger.warn("[{}] cannot be imported as a dangling index, as an index with the same name and UUID exist in the "
```
I mean the case where it doesn't get deleted synchronously and gets added to `pendingDeletes` instead.

pugnascotia replied:
I'm not very familiar with the mechanics of deletion, though I see you point about the log message and I could certainly remove it.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia committed:
Use IndexGraveyardDiff to compare graveyards

pugnascotia committed:
Change class hierarchy of ImportDanglingIndexRequest

By extending AcknowledgedRequest, it becomes possible to replace
ImportDanglingIndexResponse with AcknowledgedResponse.

pugnascotia committed:
Replace test strings with random UUIDs

pugnascotia committed:
Guard against duplicate tombstones for dangling indices

pugnascotia committed:
Tweaking

pugnascotia commented on server/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -256,18 +256,37 @@ public boolean isNewCluster() {
+        for (IndexGraveyard.Tombstone tombstone : currentMetadata.indexGraveyard().getTombstones()) {
+            final Index index = tombstone.getIndex();
+            final boolean isNewTombstone = previousMetadata.hasIndex(index) == false
+                && previousMetadata.indexGraveyard().containsIndex(index) == false;
```
Now using `IndexGraveyardDiff`.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/import_index/ImportDanglingIndexRequest.java:
```diff
@@ -0,0 +1,70 @@
+ * by its UUID. The {@link #acceptDataLoss} flag must also be
+ * explicitly set to true, or later validation will fail.
+ */
+public class ImportDanglingIndexRequest extends BaseNodesRequest<ImportDanglingIndexRequest> {
```
I swapped it for `AcknowledgedRequest`, which aligns with the delete request and meant I could remove `ImportDanglingIndexResponse` entirely.

pugnascotia commented on server/src/test/java/org/elasticsearch/action/admin/indices/dangling/list/ListDanglingIndicesResponseTests.java:
```diff
@@ -0,0 +1,147 @@
+
+        final var danglingIndexInfo = List.of(
+            new DanglingIndexInfo("node-id-1", "some-index", "deadb33f", 123456L),
+            new DanglingIndexInfo("node-id-1", "some-other-index", "cafebabe", 7891011L)
```
Now using UUIDs!

pugnascotia commented on server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java:
```diff
@@ -134,38 +152,51 @@ void cleanupAllocatedDangledIndices(Metadata metadata) {
+        // index is no longer considered dangling.
+        for (Index key : danglingIndices.keySet()) {
+            if (graveyard.containsIndex(key)) {
+                danglingIndices.remove(key);
```
Changed as you suggest.

pugnascotia commented on server/src/main/java/org/elasticsearch/action/admin/indices/dangling/delete/TransportDeleteDanglingIndexAction.java:
```diff
@@ -0,0 +1,228 @@
+        Metadata.Builder metaDataBuilder = Metadata.builder(metaData);
+
+        final IndexGraveyard newGraveyard = IndexGraveyard.builder(metaDataBuilder.indexGraveyard())
+            .addTombstone(indexToDelete)
```
I believe that in order to get to this point in the code, the index needs to be (1) missing from the metadata, and (2) missing from the graveyard. But, it doesn't hurt to have an extra check for safety, so I've added it.

pugnascotia committed:
Remove unnecessary log line

pugnascotia committed:
Test fixes

pugnascotia commented:
After talking to @DaveCTurner, I've removed the log line in question, as well as fix a test break. I believe there are now no outstanding issues on this PR.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 48366-dangling-indices-api

pugnascotia commented:
@DaveCTurner @ywelsch is there anything else you'd like to see changed on this PR?

ywelsch commented:
Once this is merged, we should also add the LIST API to the [support diagnostics](https://github.com/elastic/support-diagnostics) tool.

DaveCTurner reviewed (approved):
Thanks @pugnascotia, this LGTM :)

ywelsch reviewed (approved):
I've only checked the core changes in `IndicesClusterStateService`, `ClusterChangedEvent` and `DanglingIndicesState`, which looked good to me.
